pkg/pool-weighted/contracts/smart/ManagedPool.sol

```diff
@@ -444,6 +445,268 @@ contract ManagedPool is BaseWeightedPool, ProtocolFeeCache, ReentrancyGuard {
         emit SwapEnabledSet(swapEnabled);
     }
 
+    /**
+     * @dev This function takes the token, and the normalizedWeight it should have in the pool after being added.
+     * The stored (denormalized) weights of all other tokens remain unchanged, but the weightSum will increase,
+     * such that the normalized weight of the new token will match the target value, and the normalized weights of
+     * all other tokens will be reduced proportionately.
+     *
+     * addToken performs the following operations:
+     *
+     * - Verify there is room for the token, there is no weight change, and the final weights are all valid
+     * - Calculate the new BPT price, and ensure it is >= the minimum
+     * - Register the new token with the Vault
+     * - Join the pool, pulling in the tokenAmountIn from the sender
+     * - Increase the total weight, and update the number of tokens and `_tokenState`
+     *   (all other weights then scale accordingly)
+     * - Return the BPT value of the new token, for possible use by the caller
+     */
+    function addToken(
+        IERC20 token,
+        uint256 normalizedWeight,
+        uint256 tokenAmountIn,
+        address assetManager,
+        uint256 minBptPrice,
```

Comment:
Let's consider having a `minBptAmountOut` parameter. It feels weird to have a slippage limit set as a rate rather than a concrete amount like we do elsewhere.